### DISCUSSION TITLE
Triggering a DAG from another DAG and wait for its completion

### DISCUSSION THREAD

**Discussion by yashk97** [apache/airflow#15049] [General] [answered]
I am using Airflow 2.0.0, MySQL 5.7 on Python 3.8 for my workflow.

Consider 2 DAGs, DAG A and DAG B.

Now DAG A performs some tasks and then Triggers DAG B. DAG A is successful only if DAG B is successful i.e. A's completion depends on B.

I came across TriggerDagRunOperator as it has a wait for completion field (I did not use ExternalTaskSensor as I might have cases where DAG B will have to be retriggered if failed).

I am able to trigger DAG B from A but my conf from DAG A is not getting passed on to DAG B

```
dag_a_branch_task=BranchPythonOperator(
    task_id='dag_a_branch_task',
    python_callable=branch_task,
    on_failure_callback=post_branching_failure,
    provide_context=True,
    dag=dag_a)

trigger_dag_b = TriggerDagRunOperator(
    task_id='trigger_dag_b',
    trigger_dag_id='dag_b_task',
    execution_date='{{ ds }}',
    reset_dag_run=True,
    wait_for_completion=True,
    poke_interval=10,
    conf={"some_data_key":"{{ ti.xcom_pull(task_ids='dag_a_branch_task',key='some_data_key')}}"},
    dag=dag_a
)

```
In the dag_b_task I am using this to get the conf (as done in the case of Python Operators and Branch Operators):
`some_data` = kwargs['dag_run'].conf['some_data_key']

I get KeyError on some_data

Am I missing out on something? Any suggestions will be appreciated.

Thanks in advance!

**Comment by turbaszek** [comment]
Have you tried to add `BashOperator` or `PythonOperator` to dag b to print the conf and see it in logs just to see how it looks like? From the code perspective your code seems ok.

**Reply by yashk97** [reply]
I do have a working implementation with PythonOperator. But I need a flow where dag A waits for dag B to complete successfully.

**Reply by turbaszek** [reply]
I'm not sure what is exactly your problem. In initial comment you said that the problem is in passing the `conf` between DAGs. Now the problem seems to be in relation between tasks/dags. Could you please clarify?

**Reply by yashk97** [reply]
Sorry if I've created confusion.  My goal is to have a workflow where the status of one DAG depends on the same of a triggered DAG.

I used TriggerDagRunOperator to achieve the same because it has the wait_for_completion parameter. 

To answer your question in your first reply I did try PythonOperator and was able to get the contents of conf passed. But, correct me if I'm wrong, the PythonOperator will not wait for the completion (success/failure) of the callable python function. (I guess I forgot to mention this in my previous reply)

So to summarise I have achieved the required relationship between the DAGs using the TriggerDagRunOperator but am not able to access the conf passed in the triggered task.

**Reply by yashk97** [reply]
I think I misread your original comment.

I'll do this:
Create a sample workflow to replicate the issue about passing `conf` from `TriggerDagRunOperator` to a `PythonOperator`.
I'll try printing the `conf` and share the error logs that I am getting.

I hope this helps. Sorry for the confusing comments made earlier.

**Comment by yashk97** [comment, marked as answer]
I found a mistake in my code.
In the TriggerDagRunOperator I had passed datestamp instead of timestamp and hence it referred to the same conf (which might have been Empty) and hence gave me KeyError

`    execution_date='{{ ds }}',
`
After Changing to timestamp `'{{ ts }}'` I was able to pass `conf` to the triggered DAG. Thanks a lot and sorry for the trouble.

**Comment by lixiaoyong12** [comment]
How do you trigger DAG B after DAG A finished?

**Reply by lixiaoyong12** [reply]
And After DAG A is finished, DAG B is triggered, and DAG A continues to execute the task in the next period.
